### ISSUE TITLE
Batching rule not implemented for aten::item.

### ISSUE THREAD

**Issue by hallojs** [pytorch/functorch#337]
Hey, I would like to use `functorch.vmap` in a custom PyTorch activation function (the gradients are not needed, because the backward-pass is calculated differently). During the computation of the activation function, I do a lookup in a tensor `X` using a tensor `Y.item()` call, similar to the small dummy code below.

Unfortunately I get the error message: `RuntimeError: Batching rule not implemented for aten::item. We could not generate a fallback.`

Is it not possible to do an item() call in a vmap function or is something else wrong? Thanks a lot!

```python
import torch
from functorch import vmap

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

sum = torch.zeros([100, 10], dtype=torch.int32).to(device)
lookup = torch.randint(100, (20, 1000, 10)).to(device)
input_tensor = torch.randint(1000, (100, 20)).to(device)

def test_fun(sum, input_tensor):
  for j in range(20):
    for i in range(10):
      sum[i] += lookup[j, input_tensor[j].item(), i]
  return sum

# non-vectorized version
for i in range(100):
  test_fun(sum[i], input_tensor[I])

# vectorized version throws error
test_fun_vec = vmap(test_fun)
test_fun_vec(sum, input_tensor)
```

**Comment by Chillee** [comment]
@hallojs 

Hey, so there are 2 levels to your question :)

> Do we support `.item()`?

Essentially, no. `.item()` returns a Scalar value , which we can't vmap over.

> Can we vmap over the function you're trying to write?

We should be able to, in theory (at least, we'd want to). I think this is more of a concrete problem with slicing and such. Lemme think a bit more about it and get back to you.

**Comment by hallojs** [comment]
Hi @Chillee, thanks for your fast response!

> Essentially, no. .item() returns a Scalar value , which we can't vmap over.

Unfortunately, I don't really get it. When I vectorize the same code with JAX's `vmap` I don't have the problem. And also in this case `input_tensor[j]` is a scalar, isn't it? I thought `functorch.vmap` should work like `jax.vmap`. Or do I understand something wrong?

```python
import jax.numpy as jnp
from jax import vmap, random

key = random.PRNGKey(0)
sum = jnp.zeros([100, 10])
lookup = random.randint(key, [20, 1000, 10], 0, 100)
input_tensor = random.randint(key, [100, 20], 0, 1000)

def test_fun(sum, input_tensor):
  for j in range(20):
    for i in range(10):
      sum = sum.at[i].set(sum[i] + lookup[j, input_tensor[j], i])
  return sum

test_fun_vec = vmap(test_fun)

sum = test_fun_vec(sum, input_tensor)
```

> We should be able to, in theory (at least, we'd want to). I think this is more of a concrete problem with slicing and such. Lemme think a bit more about it and get back to you.

Thanks for your help! :)

**Comment by vfdev-5** [comment]
@hallojs the idea is to have all operations on tensors without using scalars as python int, float etc

In your functorch code: https://github.com/pytorch/functorch/issues/337#issue-1079624922, the issue we have is with `select` or `[...]` op. Inside vmap doing something like `lookup[:, input_tensor[0], :]` will implicitly try to convert `input_tensor[0]` tensor to python int and thus the error.

I think your code can be rewritten like that:
```python
import torch
from functorch import vmap

device = "cpu"
sum = torch.zeros([100, 10], device=device)
lookup = torch.randint(100, (20, 1000, 10), device=device)
input_tensor = torch.randint(1000, (100, 20), device=device)

def test_fun(sum, input_tensor):
  lookup2 = lookup.index_select(1, input_tensor)
  for j in range(20):
    for i in range(10):
      sum[i] += lookup2[j, 0, i]
  return sum

# vectorized version throws error
out = vmap(test_fun)(sum, input_tensor)
```

Hope this helps

**Comment by vfdev-5** [comment]
@hallojs I'm not sure about JAX but just tested your code snippet with jax 0.2.26 and it also tells me an error with item
```
AttributeErrorTraceback (most recent call last)
<ipython-input-3-1dc4a1b73462> in <module>
     18 test_fun_vec = vmap(test_fun)
     19 
---> 20 sum = test_fun_vec(sum, input_tensor)

    [... skipping hidden 3 frame]

<ipython-input-3-1dc4a1b73462> in test_fun(sum, input_tensor)
     10   for j in range(20):
     11     for i in range(10):
---> 12       sum = sum.at[i].set(sum[i] + lookup[j, input_tensor[j].item(), i])
     13   return sum
     14 #     v = input_tensor[0].item()

/opt/conda/lib/python3.8/site-packages/jax/core.py in __getattr__(self, name)
    572 
    573     try:
--> 574       attr = getattr(self.aval, name)
    575     except KeyError as err:
    576       raise AttributeError(

AttributeError: 'ShapedArray' object has no attribute 'item'
```

**Comment by hallojs** [comment]
@vfdev-5 Oh yes, I was a little bit too fast. The code only runs without the `.item()` call. Sorry!

**Comment by vfdev-5** [comment]
Yes, it is a bit unfortunate that [torch.select](https://pytorch.org/docs/stable/generated/torch.Tensor.select.html) and `[...]` works on ints (there should be a reason for that), otherwise functorch implementation should be exactly the same.

**Comment by hallojs** [comment]
Yes, I think that's the thing I stumbled across. Thanks again for your help! :)

**Comment by Chillee** [comment]
@hallojs functorch’s vmap works in the same way as Jax’s, but there are certain operators that have slightly different constraints in Pytorch. In this case, I think the solution is that we need to modify select to support tensors and not just integers.

**Comment by hallojs** [comment]
@Chillee All right, thanks!😊 Yes I think that would be a valuable functional extension.
